### DISCUSSION TITLE
Video thumbnail

### DISCUSSION THREAD

**Discussion by paulo-caldas** [filebrowser/filebrowser#1486] [Ideas]
**Is your feature request related to a problem? Please describe.**
The ability to see an image thumbnail is pretty useful, would enjoy something similar for video files too. Would make browsing and identifying files much easier

**Describe the solution you'd like**
Something similar to what's being done with pictures at the moment, so just take one single frame of the video and use it as the thumbnail

**Comment by niubility000** [comment]
good idea!

**Comment by cody82** [comment]
I implemented video thumbnails here: https://github.com/cody82/filebrowser/tree/videopreview
You have to install ffmpeg.
It works quite nice but does not limit ffmpeg processes yet, so it might overload your server.

**Comment by niubility000** [comment]
Just use the browser to generate the thumbnails, it's much simpler. Just add a few codes in listingitem.vue

    <div>
      <img
        v-if="readOnly == undefined && type === 'image' && isThumbsEnabled"
        v-lazy="thumbnailUrl"
      />
      <div v-else-if= "readOnly == undefined && type === 'video' && isThumbsEnabled && isMosaicMode">
	<video :src="videoUrl"></video>
        <i style="position:absolute; top:1px !important;fontSize:2em !important;color:white;" class="material-icons">videocam</i>	
      </div>  
      <i v-else class="material-icons">{{ icon }}</i>
    </div>

**Reply by thallium** [reply]
How do you implement `videoUrl`? I did it with `/api/raw` but the performance seems slow.

**Comment by niubility000** [comment]
![000759_77641ae7_9744606](https://user-images.githubusercontent.com/76441520/147891624-5d2eaa01-14c4-4c12-bf4b-7fffa7fc6e9f.png)
I use larger thumbnails personally.

**Comment by cody82** [comment]
The browser thumbnails have a few drawbacks:
- Traffic is high
- Performance is slow(mainly because of traffic and lack of caching)
- Only supports a few video formats

You cant do much about that. I would like to see official support for server side video thumbnails!

**Reply by thallium** [reply]
I ended up using your solution and it works really well.

**Reply by maxlieb** [reply]
Could you please help me find the exact code modifications you did?
I want to learn how to do this mod on the latest official version.

**Reply by cody82** [reply]
Hi, I rebased my branch on the latest master, you can view the changes here: https://github.com/filebrowser/filebrowser/compare/master...cody82:filebrowser:videopreview

**Reply by maxlieb** [reply]
Thank you very much!

**Reply by maxlieb** [reply]
@cody82 
BTW, as you said, "It works quite nice but does not limit ffmpeg processes yet, so it might overload your server."
Indeed it did... 
when I opened a folder with lots of videos, it filled up all RAM + swap + CPU usage 100% and crashed the LXC container I run it from.

I played around with the code and with some help from GPT was able to limit it. 
I'm still tweaking it but once I'm happy with the results I can publish my fork (with credits to you)

**Reply by cody82** [reply]
I also added a limit: https://github.com/filebrowser/filebrowser/commit/757df4918d331c1c145a45fee92214835ce8c4cc
I don't know if there is a better way though. I think it also delays other requests than previews.
Looking forward to see your solution!

**Reply by maxlieb** [reply]
Yes, I did it that way as well.

Other things I did:
added some code to break generation when you navigate away from the folder,
so it can go on and generate for the next folder you visit before it finished the last one.

added code to save and load the thumbnails from a .thumbnails subfolder so the thumbs are not re-generated every time filebrowser is reloaded or whatever else happens to the cache...

modified cache key function to truncate long filenames because I had an issue with too long thumbnail filenames.

added ffmpeg parameters to use intel qsv, not sure how great of impact that had but won't hurt :)

next thing I need to modify is enable those previews on a shared links page...
hint where i can find that?

**Reply by maxlieb** [reply]
https://github.com/maxlieb/filebrowser

**Comment by colelisiongaming** [comment]
Is there a more up to date implementation of this?

**Reply by maxlieb** [reply]
yes, even two :)

**Comment by maxlieb** [comment]
the repo is up to date now and will be automatically kept so from now on.
As for releases I compiled and uploaded one today, but I have not automated this yet.
you are welcome to use it or compile yourself

**Reply by JacquesVonHamsterviel** [reply]
can you compile the windows x64 version as well, thank you.

**Reply by maxlieb** [reply]
As noted in the README, my modifications are Linux-specific. 
You might run the Linux version on Windows via WSL, 
but for best performance, use a native Linux host (either direct or containerized with GPU passthrough) 
and Intel integrated graphics for hardware acceleration.

**Reply by maxlieb** [reply]
Last few days I have put in work on the project and made it cross platform with HW acceleration working on windows and theoretically mac as well, you are welcome to try.
Credits to @shengj1ang for portions of code for GitHub actions and video player controls fix.

**Comment by shengj1ang** [comment]
I use GitHub Actions to compile the version that support video thumbnails. Widely support Windows, MacOS, Linux x86 x64.
and also fix: set videojs locale https://github.com/filebrowser/filebrowser/pull/3742

ffmpeg is required

https://github.com/shengj1ang/filebrowser/releases/tag/202502240418
